### PR TITLE
[ICD] Send Check-In messages

### LINKED ISSUE #24243: [ICD] Implement Check-In Message sending in the ICDManager
### Feature description
ICDManager is responsible of sending the ICD Check-In Message when clients are registered.
Implement the check-in message creation and sending

### LINKED ISSUE #28229: [ICD] Implement the ICD Check-In Message state-machine
#### Description
when an ICD goes from IdleMode to ActiveMode, it needs to send a check-in message to all registered controllers if they do not have a valid subscription with the ICD.

The steps are :
1- For each registered subject, check if they have an active subscription 
2- If not, send a check-in message to the check-in node Id.

image

### LINKED ISSUE #30492: [1.3][ICD] Stay active during MDNS resolution for sending checking messages
When resolving NodeID, the device needs to stay active.

### PR DESCRIPTION
**Pull request by jepenven-silabs**
Fix #28229 
Fix #24243 

Add Logic to trigger sending of the ICD Check-In message

### CODE REVIEW

File: src/app/icd/ICDManager.h [hunk 1]
```diff
@@ -154,6 +158,8 @@ class ICDManager : public ICDListener
     // feature map that can be changed at runtime for testing purposes
     uint32_t mFeatureMap = 0;
 #endif
+
+    static ObjectPool<ICDCheckInSender, (CHIP_CONFIG_ICD_CLIENTS_SUPPORTED_PER_FABRIC * CHIP_CONFIG_MAX_FABRICS)> mICDSenderPool;
```
Comment (jmartinez-silabs) [inline src/app/icd/ICDManager.h:162]: Does the pool have to be static? Can it not be created at init and teardown at Shutdown?

**Review by jmartinez-silabs — APPROVED** [review]

File: src/platform/telink/BUILD.gn [hunk 2]
```diff
@@ -58,6 +58,10 @@ static_library("telink") {
   public_deps = [ "${chip_root}/src/platform:platform_base" ]
   deps = [ "${chip_root}/src/platform/logging:headers" ]
 
+  public_configs = [
```
Comment (andy31415) [inline src/platform/telink/BUILD.gn:61]: I think the linux need for this even on Linux was questionable.
Could you please explain why this is needed?

**Review by andy31415 — CHANGES_REQUESTED** [review]
Request changes: I am not sure we have good dependency tracking and adding more `default_address_resolve_config` dependencies at a platform level is concerning.

File: src/app/icd/ICDManager.cpp [hunk 3]
```diff
@@ -41,9 +44,12 @@ using namespace chip::app::Clusters;
 using namespace chip::app::Clusters::IcdManagement;
 
 uint8_t ICDManager::OpenExchangeContextCount = 0;
+uint8_t ICDManager::CheckInRequestCount      = 0;
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:47]: these don't need to be static anymore.

File: src/app/icd/ICDManager.cpp [hunk 4]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
     return success ? ((featureMap & to_underlying(feature)) != 0) : false;
 #else
-
     return ((mFeatureMap & to_underlying(feature)) != 0);
 #endif // !CONFIG_BUILD_FOR_HOST_UNIT_TEST
 }
 
+void ICDManager::SendCheckInMsgs()
+{
+#if !CONFIG_BUILD_FOR_HOST_UNIT_TEST
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:113]: Is this behind the define because of the interactione model? 
This is a problem because we will want to unit-test this.

File: src/app/icd/ICDManager.cpp [hunk 5]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+    VerifyOrDie(mFabricTable != nullptr);
+    for (const auto & fabricInfo : *mFabricTable)
+    {
+        uint16_t supported_clients = ICDManagementServer::GetInstance().GetClientsSupportedPerFabric();
+
+        ICDMonitoringTable table(*mStorage, fabricInfo.GetFabricIndex(), supported_clients /*Table entry limit*/,
+                                 mSymmetricKeystore);
+        if (!table.IsEmpty())
+        {
+            Messaging::ExchangeManager * exchangeManager = InteractionModelEngine::GetInstance()->GetExchangeManager();
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:124]: ExchangeManager is owned by the Server. This can be given when the ICDManager is initialized.

File: src/app/icd/ICDManager.cpp [hunk 6]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+
+            for (uint16_t i = 0; i < table.Limit(); i++)
+            {
+                ICDMonitoringEntry entry(mSymmetricKeystore);
+                if (CHIP_NO_ERROR != table.Get(i, entry))
+                {
+                    // No more entries in the table
+                    break;
+                }
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:133]: `table.Get(...)` can returns errors that do not indicate that the table is empty.
We need to check the specific action a take the appropriate action.

File: src/app/icd/ICDManager.cpp [hunk 7]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                    continue;
+                }
+
+                ICDCheckInSender * sender = mICDSenderPool.CreateObject(exchangeManager);
+
+                if (sender == nullptr)
+                {
+                    ChipLogError(AppServer, "Failed to allocate ICDCheckinSender");
+                    return;
+                }
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:149]: can be simplified with a VerifyOrReturn.

File: src/app/icd/ICDManager.cpp [hunk 8]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+
+                if (CHIP_NO_ERROR != sender->RequestResolve(entry, mFabricTable))
+                {
+                    ChipLogError(AppServer, "Failed to send ICD Check-In");
+                }
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:154]: can be simplified with a VerifyOrReturn.

File: src/app/icd/ICDManager.cpp [hunk 9]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                bool active =
+                    InteractionModelEngine::GetInstance()->IsMatchingSubscriptionActive(entry.fabricIndex, entry.monitoredSubject);
+
+                if (active)
+                {
+                    continue;
+                }
+
+                ICDCheckInSender * sender = mICDSenderPool.CreateObject(exchangeManager);
+
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:144]:
```suggestion
```

File: src/app/icd/ICDManager.cpp [hunk 10]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                ICDMonitoringEntry entry(mSymmetricKeystore);
+                if (CHIP_NO_ERROR != table.Get(i, entry))
+                {
+                    // No more entries in the table
+                    break;
+                }
+
+                bool active =
+                    InteractionModelEngine::GetInstance()->IsMatchingSubscriptionActive(entry.fabricIndex, entry.monitoredSubject);
+
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:137]:
```suggestion
```

File: src/app/icd/ICDManager.cpp [hunk 11]
```diff
@@ -168,10 +223,13 @@ void ICDManager::UpdateOperationState(OperationalState state)
...
 
+        // Going back to Idle, all check-in messages are sent
+        mICDSenderPool.ReleaseAll();
+
         CHIP_ERROR err = DeviceLayer::ConnectivityMgr().SetPollingInterval(slowPollInterval);
-        if (err != CHIP_NO_ERROR)
+        if (err != CHIP_NO_ERROR && err != CHIP_ERROR_NOT_IMPLEMENTED)
         {
-            ChipLogError(AppServer, "Failed to set Polling Interval: err %" CHIP_ERROR_FORMAT, err.Format());
+            ChipLogError(AppServer, "Failed to set Slow PollinFg Interval: err %" CHIP_ERROR_FORMAT, err.Format());
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:232]:
```suggestion
            ChipLogError(AppServer, "Failed to set Slow Polling Interval: err %" CHIP_ERROR_FORMAT, err.Format());
```

File: src/app/icd/ICDManager.cpp [hunk 12]
```diff
@@ -168,10 +223,13 @@ void ICDManager::UpdateOperationState(OperationalState state)
         }
 #endif
 
+        // Going back to Idle, all check-in messages are sent
+        mICDSenderPool.ReleaseAll();
+
         CHIP_ERROR err = DeviceLayer::ConnectivityMgr().SetPollingInterval(slowPollInterval);
-        if (err != CHIP_NO_ERROR)
+        if (err != CHIP_NO_ERROR && err != CHIP_ERROR_NOT_IMPLEMENTED)
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:230]: Not sure we want to suppress the not implemented error...

File: src/app/icd/ICDManager.cpp [hunk 13]
```diff
@@ -199,9 +257,14 @@ void ICDManager::UpdateOperationState(OperationalState state)
             DeviceLayer::SystemLayer().StartTimer(System::Clock::Timeout(activeModeJitterInterval), OnTransitionToIdle, this);
 
             CHIP_ERROR err = DeviceLayer::ConnectivityMgr().SetPollingInterval(GetFastPollingInterval());
-            if (err != CHIP_NO_ERROR)
+            if (err != CHIP_NO_ERROR && err != CHIP_ERROR_NOT_IMPLEMENTED)
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:260]: same comment

File: src/app/icd/ICDManager.h [hunk 14]
```diff
@@ -128,6 +131,7 @@ class ICDManager : public ICDListener
     static void OnTransitionToIdle(System::Layer * aLayer, void * appState);
 
     static uint8_t OpenExchangeContextCount;
+    static uint8_t CheckInRequestCount;
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.h:134]: don't need to be static anymore

File: src/app/server/Server.cpp [hunk 15]
```diff
@@ -328,6 +319,15 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
                                                                  &mCASESessionManager, mSubscriptionResumptionStorage);
     SuccessOrExit(err);
 
+    // ICD Init needs to be after data model init and InteractionModel Init
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+    mICDManager.Init(mDeviceStorage, &GetFabricTable(), mSessionKeystore);
```
Comment (mkardous-silabs) [inline src/app/server/Server.cpp:324]: add the mExchangeMgr to the past pointers.

File: src/app/InteractionModelEngine.cpp [hunk 16]
```diff
@@ -278,6 +278,35 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
     return CHIP_ERROR_KEY_NOT_FOUND;
 }
 
+bool InteractionModelEngine::IsMatchingSubscriptionActive(const FabricIndex & aFabricIndex, const NodeId & aPeerNodeId)
+{
+    // This is assuming that ReadClient::Close will not affect any other
+    // ReadClients in the list.
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:285]: This doesn't make any sense.  Shouldn't this be looking at read _handlers_, not read _clients_?  Was this tested at all?

File: src/app/icd/ICDManager.cpp [hunk 17]
```diff
@@ -98,11 +104,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+
+                if (sender == nullptr)
+                {
+                    ChipLogError(AppServer, "Failed to allocate ICDCheckinSender");
+                    return;
+                }
+
+                if (CHIP_NO_ERROR != sender->RequestResolve(entry, mFabricTable))
+                {
+                    ChipLogError(AppServer, "Failed to send ICD Check-In");
```
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:153]: Doesn't this leak the sender?

File: src/app/icd/ICDManager.cpp [hunk 18]
```diff
@@ -168,10 +223,13 @@ void ICDManager::UpdateOperationState(OperationalState state)
         }
 #endif
 
+        // Going back to Idle, all check-in messages are sent
+        mICDSenderPool.ReleaseAll();
```
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:227]: There should be documentation around where the allocation from the pool happens that makes it clear how the lifetimes are managed here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/icd/ICDManager.cpp (hunk 17 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:153]: No since the pool is release at the next Idle mode transition

File: src/app/icd/ICDManager.cpp (hunk 11 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:232]: Changed on PR https://github.com/project-chip/connectedhomeip/pull/30552

File: src/app/icd/ICDManager.cpp (hunk 13 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:260]: Changed on PR https://github.com/project-chip/connectedhomeip/pull/30552

File: src/app/icd/ICDManager.cpp (hunk 12 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:230]: Changed on PR https://github.com/project-chip/connectedhomeip/pull/30552

File: src/app/InteractionModelEngine.h [hunk 19]
```diff
@@ -306,6 +306,11 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
 
     CHIP_ERROR ResumeSubscriptions();
 
+    /**
+     * Helper to find the state of a matching subscription
+     */
```
[the remark this answers is not in the document]
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.h:311]: removed

File: src/app/InteractionModelEngine.cpp [hunk 20]
```diff
@@ -278,6 +278,35 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
...
+    {
+        // Grab the next client now, because we might be about to delete readClient.
+        auto * nextClient = readClient->GetNextClient();
+        if (readClient->IsSubscriptionType())
+        {
+            if (IsCASEAuthTag(aPeerNodeId))
+            {
+                // TODO ReadClient doesn't seems to hold CASEAuthTag
```
[the remark this answers is not in the document]
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:293]: #30553

File: src/app/InteractionModelEngine.cpp [hunk 21]
```diff
@@ -278,6 +278,35 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
     return CHIP_ERROR_KEY_NOT_FOUND;
 }
 
+bool InteractionModelEngine::IsMatchingSubscriptionActive(const FabricIndex & aFabricIndex, const NodeId & aPeerNodeId)
+{
+    // This is assuming that ReadClient::Close will not affect any other
+    // ReadClients in the list.
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
+    {
+        // Grab the next client now, because we might be about to delete readClient.
```
[the remark this answers is not in the document]
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:287]: removed

File: src/app/icd/ICDManager.cpp (hunk 4 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:113]: This is not unit-testable because we do need the interaction model. MoreOver, this triggers a MDNS resolve which is not UnitTestable. This part will have to be covered by Integration test only.

File: src/app/icd/ICDManager.cpp (hunk 6 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:133]: Is you consult the implementation of the ICD Monitoring Table (I guess you've already done so since you've reviewed the PR #26809 ) This is the only way to iterate through all elements of the table. No action needs to be done upon failure of the `Get` command since what the `Get` command do is to load something from persistent storage. It's either there or not.

File: src/app/icd/ICDManager.cpp (hunk 8 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:154]: Not here since we do want to iterate through all the ICD Monitoring entry and no simply exit the function directly.

File: src/app/icd/ICDManager.cpp (hunk 6 above)
Comment (mkardous-silabs) [reply src/app/icd/ICDManager.cpp:129]: ICDMonitoringTable, in the `IsEmpty` function explicitly checks that the `Get` function returns `CHIP_ERROR_NOT_FOUND` to determine if the table is empty or not. We should the same.

```
1. if table.Get == CHIP_ERROR_NOT_FOUND  -> table is empty we can break
2. if table.Get == CHIP_NO_ERROR -> Continue
3. if table.Get != CHIP_NO_ERROR && table.Get != CHIP_ERROR_NOT_FOUND -> Error occured but the table might not be empty
```
For the third case, we need to device if we check the next entry or error out of the whole process.

File: src/app/icd/ICDManager.cpp (hunk 4 above)
Comment (mkardous-silabs) [reply src/app/icd/ICDManager.cpp:109]: Can you create a follow issue for this? It would be worth while to investigate how the CASE test work.

File: src/app/icd/ICDManager.cpp (hunk 17 above)
Comment (bzbarsky-apple) [reply src/app/icd/ICDManager.cpp:147]: What if there is no next transition?

File: examples/lit-icd-app/linux/config/CHIPProjectConfig.h
[the remark this answers is not in the document]
Comment (jepenven-silabs) [reply examples/lit-icd-app/linux/config/CHIPProjectConfig.h:1]: #30616 30616

File: src/app/icd/ICDManager.cpp (hunk 4 above)
Comment (mkardous-silabs) [reply src/app/icd/ICDManager.cpp:112]: Was an issue created for this?

File: src/app/icd/ICDManager.cpp [hunk 22]
```diff
@@ -101,6 +107,60 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                }
+
+                // SenderPool will be release upon transition from active to idle state
+                // This will happen when all ICD Check-In messages are sent on the network
+                ICDCheckInSender * sender = mICDSenderPool.CreateObject(mExchangeManager);
+                if (sender == nullptr)
+                {
+                    ChipLogError(AppServer, "Failed to allocate ICDCheckinSender");
+                    return;
+                }
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:152]:
```suggestion
VerifyOrReturn(sender != nullptr, ChipLogError(AppServer, "Failed to allocate ICDCheckinSender");
```

File: src/app/icd/ICDManager.cpp [hunk 23]
```diff
@@ -166,6 +226,9 @@ void ICDManager::UpdateOperationState(OperationalState state)
         }
 #endif
 
+        // Going back to Idle, all check-in messages are sent
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:229]:
```suggestion
        // Going back to Idle, all Check-In messages have been sent
```

File: src/app/InteractionModelEngine.h [hunk 24]
```diff
@@ -306,6 +306,9 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
 
     CHIP_ERROR ResumeSubscriptions();
 
+    // Return the state of a subscription matching a CAT Tag or a NodeId
+    bool IsMatchingSubscriptionActive(const FabricIndex & aFabricIndex, const NodeId & nodeIdOrCAT);
```
Comment (mkardous-silabs) [inline src/app/InteractionModelEngine.h:310]: naming isn't clear with what the function does; we aren't matching.
maybe `doesSubjectHaveActiveSubscription`? or a verison that with the associated comment.

Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.h:310]: Well to me we are indeed Matching the subscription against the fabricId AND (NodeId OR CAT). 

What would be a Match if not the above ?

Will make the comment/function name clearer

File: src/app/icd/ICDManager.cpp (hunk 4 above)
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:112]: #30658

**Review by mkardous-silabs — APPROVED** [review]

File: src/app/icd/ICDManager.cpp [hunk 25]
```diff
@@ -101,6 +107,56 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+    VerifyOrDie(mStorage != nullptr);
+    VerifyOrDie(mFabricTable != nullptr);
+    for (const auto & fabricInfo : *mFabricTable)
+    {
+        uint16_t supported_clients = ICDManagementServer::GetInstance().GetClientsSupportedPerFabric();
+
+        ICDMonitoringTable table(*mStorage, fabricInfo.GetFabricIndex(), supported_clients /*Table entry limit*/,
+                                 mSymmetricKeystore);
+        if (!table.IsEmpty())
+        {
```
Comment (andy31415) [inline src/app/icd/ICDManager.cpp:122]: is this extra check worth the optimization indent? Otherwise I would expect we could decrease method indent by just leaving the for loop. 

Or if needed something like:

```
if (table.IsEmpty())
{
   // avoid a loop over all items if all would return NOT_FOUND
   continue;
}
```

**Review by andy31415 — APPROVED** [review]

File: src/app/icd/ICDManager.cpp [hunk 26]
```diff
@@ -101,6 +107,56 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                ICDMonitoringEntry entry(mSymmetricKeystore);
+                CHIP_ERROR err = table.Get(i, entry);
+                if (err == CHIP_ERROR_NOT_FOUND)
+                {
+                    break;
+                }
+
+                if (err != CHIP_NO_ERROR)
+                {
+                    // Try to fetch the next entry.
```
[the remark this answers is not in the document]
Comment (jepenven-silabs) [reply src/app/icd/ICDManager.cpp:134]: This was put in place to address this comment https://github.com/project-chip/connectedhomeip/pull/30526#discussion_r1397719592.

File: examples/lit-icd-app/linux/config/CHIPProjectConfig.h [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+#endif
+
+#ifndef CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION_STRING
+#define CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION_STRING "1.0"
+#endif
+
+//
+// Default of 8 ECs is not sufficient for some of the unit tests
+// that try to validate multiple simultaneous interactions.
+// In tests like TestReadHandler_MultipleSubscriptions, we are trying to issue as many read / subscription requests as possible in
```
Comment (bzbarsky-apple) [inline examples/lit-icd-app/linux/config/CHIPProjectConfig.h:80]: In what way is this config file relevant to TestReadHandler_MultipleSubscriptions ?

File: src/app/InteractionModelEngine.cpp [hunk 28]
```diff
@@ -278,6 +278,37 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
     return CHIP_ERROR_KEY_NOT_FOUND;
 }
 
+bool InteractionModelEngine::IsSubjectSubscriptionActive(const FabricIndex & aFabricIndex, const NodeId & subjectID)
+{
+    for (uint16_t index = 0; index < GetNumActiveReadHandlers(); index++)
+    {
+        ReadHandler * readHandler = ActiveHandlerAt(index);
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:285]: This is an O(N^2) traversal of the list, no?  Why is this not doing `mReadHandlers.ForEachActiveObject` directly like the other things in this file that want to walk that list?

File: src/app/InteractionModelEngine.cpp [hunk 29]
```diff
@@ -278,6 +278,37 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
...
+        if (readHandler == nullptr)
+        {
+            // Should not happen
+            continue;
+        }
+
+        if (readHandler->IsType(ReadHandler::InteractionType::Subscribe))
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:292]: I would suggest reversing the test, returning/continuing early if not a subscribe handler, and outdenting the actual code.

File: src/app/InteractionModelEngine.cpp [hunk 30]
```diff
@@ -278,6 +278,37 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
...
+            if (subject.fabricIndex != aFabricIndex)
+            {
+                continue;
+            }
+
+            if (subject.authMode == Access::AuthMode::kCase)
+            {
+                if (subject.cats.CheckSubjectAgainstCATs(subjectID) || subjectID == subject.subject)
+                {
+                    return readHandler->IsActiveSubscription();
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:304]: Why this specific boolean?  If we're in the middle of setting up a subscription with this client, wouldn't we still want to treat that as "subscription exists"?

File: src/app/InteractionModelEngine.h [hunk 31]
```diff
@@ -306,6 +306,9 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
 
     CHIP_ERROR ResumeSubscriptions();
 
+    // Return the state of a subscription for a given subject (CAT or NodeId)
+    bool IsSubjectSubscriptionActive(const FabricIndex & aFabricIndex, const NodeId & subject);
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:310]: Depending on the answer to the "what about the case when we are in the middle of setting up that subscription", this API's name might need to change.

File: src/app/icd/ICDManager.cpp [hunk 32]
```diff
@@ -101,6 +107,56 @@ bool ICDManager::SupportsFeature(Feature feature)
...
+                }
+
+                bool active =
+                    InteractionModelEngine::GetInstance()->IsSubjectSubscriptionActive(entry.fabricIndex, entry.monitoredSubject);
+                if (active)
+                {
+                    continue;
+                }
+
+                // SenderPool will be release upon transition from active to idle state
```
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:145]:
```suggestion
                // SenderPool will be released upon transition from active to idle state
```

File: src/app/InteractionModelEngine.cpp (hunk 30 above)
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:304]: In the unlikely event that this scenario happens, I would side with the fact that at the moment of the check, the subscription is not set yet. For the ICD use case, according to spec we need to check if the subscription is active or not in order to send the ICD Check-In. 

If we take the ICD use case, This check will be perform right after the device transition from Idle to active, it would be pretty unlikely that a subscription was being set-up and the device decided to go into Idle right in the middle of the subscription establishment.. Actually if that happens it should be flag as a bug and fixed.

File: examples/lit-icd-app/linux/config/CHIPProjectConfig.h (hunk 27 above)
Comment (jepenven-silabs) [reply examples/lit-icd-app/linux/config/CHIPProjectConfig.h:80]: Copy pasted from default Linux config, an issue is open to clean up this file #30616

File: src/app/InteractionModelEngine.cpp (hunk 29 above)
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:292]: done

File: src/app/InteractionModelEngine.cpp (hunk 28 above)
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:285]: done

File: examples/lit-icd-app/linux/config/CHIPProjectConfig.h (hunk 27 above)
Comment (bzbarsky-apple) [reply examples/lit-icd-app/linux/config/CHIPProjectConfig.h:80]: Can you please add comments to the file saying that most of it is nonsense and people should not copy it?  Same for the SystemProjectConfig file...

File: src/app/InteractionModelEngine.cpp [hunk 33]
```diff
@@ -318,6 +318,36 @@ void InteractionModelEngine::ShutdownMatchingSubscriptions(const Optional<Fabric
...
+        {
+            return Loop::Continue;
+        }
+
+        if (subject.authMode == Access::AuthMode::kCase)
+        {
+            if (subject.cats.CheckSubjectAgainstCATs(subjectID) || subjectID == subject.subject)
+            {
+                isActive = handler->IsActiveSubscription();
+                return Loop::Break;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:341]: OK, but what if we _do_ have an active subscription (a different one) from some other node that matches the given subject?

That is, it feels like we should only break out of the loop once we set isActive to true or run out of things to check.

File: src/app/InteractionModelEngine.cpp (hunk 30 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp]: I'm not sure I'm convinced this is the right behavior, but I agree that transitioning out of idle while in the middle of an active exchange should not happen so it does not matter too much in practice....

File: src/app/InteractionModelEngine.h (hunk 31 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.h:310]: So this should be called `SubjectHasActiveSubscription` or something.

Also, why is aFabricIndex not passed by value?  It's an 8-bit integer....  Passing the NodeId by reference maybe makes sense (or not), depending on calling conventions for 64-bit ints, but the 8-bit thing should just be passed by value.

File: src/app/InteractionModelEngine.cpp (hunk 33 above)
Comment (jepenven-silabs) [reply src/app/InteractionModelEngine.cpp:341]: Makes sense

**Review by bzbarsky-apple — APPROVED** [review]
